### PR TITLE
executor, planner: fix clustered index UnionScan and IndexLookUp

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve?

fix clustered index union scan and index lookup

### What is changed and how it works?

- add pkColIDs in MemReader and use it to decode common handle.
- fix IndexReader schema.

### Check List 

Tests 

- Unit test

### Release note 

- fix clustered index union scan and index lookup

### CODE REVIEW

**Comment by coocood** [comment]
Also fixes https://github.com/pingcap/tidb/issues/18366

File: executor/clustered_index_test.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	tk.MustExec("create table t (a int, b int, c int, d int, primary key (a, b))")
+	tk.MustExec("create index idx on t(c)")
+	tk.MustExec("insert t values (1, 1, 1, 1)")
+	tk.MustQuery("select d from t use index (idx)").Check(testkit.Rows("1"))
+}
+
+func (s *testClusteredSuite) TestClusteredIndexLookUp2(c *C) {
+	tk := s.newTK(c)
+	tk.MustExec("drop table if exists c3")
+	tk.MustExec("set @@tidb_enable_clustered_index = 1")
```
Comment (tangenta) [inline executor/clustered_index_test.go:63]:
```suggestion
```

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -988,6 +980,14 @@ func (is *PhysicalIndexScan) initSchema(idxExprCols []*expression.Column, isDoub
 			})
 		}
 	}
+	is.NeedCommonHandle = is.Table.IsCommonHandle
+
+	if is.NeedCommonHandle {
+		for i := len(is.Index.Columns); i < len(idxExprCols); i++ {
+			indexCols = append(indexCols, idxExprCols[i])
+		}
+		is.SetSchema(expression.NewSchema(indexCols...))
```
Comment (tangenta) [inline planner/core/find_best_task.go:989]:
```suggestion
		is.SetSchema(expression.NewSchema(indexCols...))
		return
```

File: planner/core/resolve_indices.go [hunk 3]
```diff
@@ -296,7 +296,7 @@ func (p *PhysicalIndexLookUpReader) ResolveIndices() (err error) {
 		p.ExtraHandleCol = newCol.(*expression.Column)
 	}
 	for i, commonHandleCol := range p.CommonHandleCols {
-		newCol, err := commonHandleCol.ResolveIndices(p.indexPlan.Schema())
+		newCol, err := commonHandleCol.ResolveIndices(p.tablePlan.Schema())
```
Comment (tangenta) [inline planner/core/resolve_indices.go:299]: Does `tablePlan.Schema()` necessarily contain common handle columns?

Comment (coocood) [reply planner/core/resolve_indices.go:299]: Yes. we never prune handle columns

**Pull request merged**